crates/llm-ls/src/adaptors.rs

```diff
@@ -0,0 +1,163 @@
+                return Err(internal_error(
+                    "TGI parser unexpectedly encountered api-inference",
+                ))
+            }
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_api_body(prompt: String, params: CompletionParams) -> Value {
+    build_tgi_body(prompt, params)
+}
+
+fn build_api_headers(api_token: Option<&String>, ide: Ide) -> Result<HeaderMap, jsonrpc::Error> {
+    build_tgi_headers(api_token, ide)
+}
+
+fn parse_api_text(text: reqwest::Result<String>) -> Result<Vec<Generation>, jsonrpc::Error> {
+    let generations =
+        match serde_json::from_str(&text.unwrap_or_default()).map_err(internal_error)? {
+            APIResponse::Generation(gen) => vec![gen],
+            APIResponse::Generations(gens) => gens,
+            APIResponse::Error(err) => return Err(internal_error(err)),
+        };
+    Ok(generations)
+}
+
+fn build_ollama_body(prompt: String, params: CompletionParams) -> Value {
+    let request_body = params.request_body.unwrap_or_default();
+    let body = serde_json::json!({
+        "prompt": prompt,
+        "model": request_body.get("model"),
+    });
+    body
+}
+fn build_ollama_headers() -> Result<HeaderMap, jsonrpc::Error> {
+    let headers = HeaderMap::new();
+    Ok(headers)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct OllamaGeneration {
+    response: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum OllamaAPIResponse {
+    Generation(OllamaGeneration),
+    Error(APIError),
+}
+
+fn parse_ollama_text(
+    text: Result<String, reqwest::Error>,
+) -> Result<Vec<Generation>, jsonrpc::Error> {
+    match text {
+        Ok(text) => {
+            let mut gen: Vec<Generation> = Vec::new();
+            for row in text.split('\n') {
+                if row.is_empty() {
+                    continue;
+                }
+                let chunk = match serde_json::from_str(row) {
+                    Ok(OllamaAPIResponse::Generation(ollama_gen)) => ollama_gen.response,
+                    Ok(OllamaAPIResponse::Error(err)) => return Err(internal_error(err)),
+                    Err(err) => return Err(internal_error(err)),
+                };
+                gen.push(Generation {
+                    generated_text: chunk,
+                })
+            }
+            Ok(gen)
+        }
+        Err(err) => Err(internal_error(err)),
+    }
+}
+
+const TGI: &str = "tgi";
+const HUGGING_FACE: &str = "huggingface";
+const OLLAMA: &str = "ollama";
+const DEFAULT_ADAPTOR: &str = HUGGING_FACE;
+
+fn unknown_adaptor_error(adaptor: String) -> jsonrpc::Error {
+    internal_error(format!("Unknown adaptor {}", adaptor))
+}
+
+pub fn adapt_body(prompt: String, params: CompletionParams) -> Result<Value, jsonrpc::Error> {
+    let adaptor = params
+        .adaptor
+        .clone()
+        .unwrap_or(DEFAULT_ADAPTOR.to_string());
+    match adaptor.as_str() {
```

Comment:
```suggestion
    match params.adaptor.unwrap_or(DEFAULT_ADAPTOR) {
```
You may need to add an `.as_ref()`, not sure

Reply:
Thanks, I'm able to get around cloning here now, but I still seem to have to do `&DEFAULT_ADAPTOR.so_string()` to string here, since `adaptor` is an `Option<&String>`